File: scripts/content-diff.py
#!/usr/bin/env python3
# Of the outputs that were rebuilt (drvPath moved vs base), which actually
# changed content? Poor man's CA early-cutoff report, run after the build:
#
#   narHash(old) == narHash(new)          -> identical, no download needed
#   no self-references on either side     -> narHash differs means changed
#   self-references                       -> fetch both, normalize with
#                                            `nix store make-content-addressed`,
#                                            compare the rewritten paths
#
# "identical" is the strong signal. "changed" may just be a rebuilt dep's
# store path embedded in an otherwise-equal output; real CA derivations would
# cut that off, but they are off the table for now (docs/architecture.md).
#
# Old outputs come from the cache (present when the base commit's CI run
# pushed them); new ones from the local store, with a cache fallback for
# builds skipped by --skip-cached.

import argparse
import json
import subprocess
import sys

CACHE_URL = "https://nix-cache.wasix.org"
CACHE_PUB_KEY = "wasinix-1:jvsqbOJGsZxMvg97fuyNCWCc+t2nn6uHB47kQCGNmXI="
NARINFO_CHUNK = 100
# make-content-addressed downloads and copies both sides; keep that bounded
MAX_NORMALIZE = 25
NAR_SIZE_CAP = 256 * 1024 * 1024
LIST_CAP = 250


def log(msg):
    print(msg, file=sys.stderr)


def path_infos(paths, store=None):
    # -> {basename: narinfo|None}; chunked, tolerant of a failing invocation
    infos = {}
    for i in range(0, len(paths), NARINFO_CHUNK):
        chunk = paths[i : i + NARINFO_CHUNK]
        cmd = ["nix", "path-info", "--json", "--json-format", "2"]
        if store:
            cmd += ["--store", store]
        p = subprocess.run(cmd + chunk, capture_output=True, text=True)
        if p.returncode != 0:
            log(f"WARN: path-info failed for a chunk: {p.stderr.strip()[:200]}")
            infos.update({basename(pth): None for pth in chunk})
            continue
        infos.update(json.loads(p.stdout)["info"])
    return infos


def basename(path):
    return path.rsplit("/", 1)[-1]


def self_referential(path, info):
    return basename(path) in info.get("references", [])


def normalize_pair(old, new):
    # -> (identical, reason|None). Realise both (old substitutes from the
    # cache), then compare with self-references rewritten to content hashes.
    # explicit cache: raw nix-store does not read the flake's nixConfig
    r = subprocess.run(
        ["nix-store", "--realise", old, new,
         "--option", "extra-substituters", CACHE_URL,
         "--option", "extra-trusted-public-keys", CACHE_PUB_KEY],
        capture_output=True, text=True,
    )
    if r.returncode != 0:
        return None, f"realise failed: {r.stderr.strip().splitlines()[-1][:120]}"
    p = subprocess.run(
        ["nix", "store", "make-content-addressed", "--json", old, new],
        capture_output=True, text=True,
    )
    if p.returncode != 0:
        return None, f"normalize failed: {p.stderr.strip().splitlines()[-1][:120]}"
    rw = json.loads(p.stdout)["rewrites"]
    return rw[old] == rw[new], None


def compare_all(pairs):
    olds = path_infos([p["old"] for p in pairs], store=CACHE_URL)
    news_local = path_infos([p["new"] for p in pairs])
    # --skip-cached builds may exist only in the cache, not locally
    missing = [p["new"] for p in pairs if not news_local.get(basename(p["new"]))]
    news_remote = path_infos(missing, store=CACHE_URL) if missing else {}

    identical, changed, skipped = [], [], []
    normalized = 0
    for p in pairs:
        old_info = olds.get(basename(p["old"]))
        new_info = news_local.get(basename(p["new"])) or news_remote.get(
            basename(p["new"])
        )
        if p["old"] == p["new"]:
            identical.append(p)
        elif old_info is None or new_info is None:
            side = "base" if old_info is None else "new"
            skipped.append((p, f"{side} output not available"))
        elif old_info["narHash"] == new_info["narHash"]:
            identical.append(p)
        elif not (
            self_referential(p["old"], old_info)
            or self_referential(p["new"], new_info)
        ):
            changed.append(p)
        elif max(old_info["narSize"], new_info["narSize"]) > NAR_SIZE_CAP:
            skipped.append((p, "self-referential and too large to normalize"))
        elif normalized >= MAX_NORMALIZE:
            skipped.append((p, f"normalize cap ({MAX_NORMALIZE}) reached"))
        else:
            normalized += 1
            same, reason = normalize_pair(p["old"], p["new"])
            if same is None:
                skipped.append((p, reason))
            elif same:
                identical.append(p)
            else:
                changed.append(p)
    return identical, changed, skipped


def section(title, lines, open_=False):
    if not lines:
        return ""
    shown = lines[:LIST_CAP]
    body = "\n".join(shown)
    if len(lines) > len(shown):
        body += f"\n- ... and {len(lines) - len(shown)} more"
    o = " open" if open_ else ""
    return f"\n<details{o}><summary>{title} ({len(lines)})</summary>\n\n{body}\n\n</details>\n"


def label(p):
    out = f"^{p['output']}" if p["output"] != "out" else ""
    return f"`{p['attr']}{out}`"


def render(pairs, identical, changed, skipped):
    md = "### Content diff\n\n"
    md += (
        f"Of **{len(pairs)}** rebuilt outputs: **{len(identical)}** bit-identical"
        f" · **{len(changed)}** changed · {len(skipped)} not comparable\n\n"
        "Identical is definitive. Changed may only be rebuilt store paths"
        " embedded in an otherwise-equal output.\n"
    )
    md += section("Changed", [f"- {label(p)}" for p in changed], open_=len(changed) <= 20)
    md += section("Bit-identical", [f"- {label(p)}" for p in identical])
    md += section("Not comparable", [f"- {label(p)}: {r}" for p, r in skipped])
    return md


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--base-map", required=True)
    ap.add_argument("--head-map", required=True)
    ap.add_argument("--md-out", required=True)
    ap.add_argument("--summary-out", required=True)
    args = ap.parse_args()

    def done(md, summary):
        with open(args.md_out, "w") as f:
            f.write(md)
        with open(args.summary_out, "w") as f:
            json.dump(summary, f)

    try:
        with open(args.base_map) as f:
            base = json.load(f)
    except OSError:
        return done("### Content diff\n\nSkipped: no base eval map.\n", {})
    with open(args.head_map) as f:
        head = json.load(f)

    rebuilt = sorted(
        a
        for a in head["jobs"].keys() & base["jobs"].keys()
        if head["jobs"][a] != base["jobs"][a]
    )
    pairs = [
        {"attr": a, "output": name, "old": old, "new": new}
        for a in rebuilt
        for name, new in head.get("outputs", {}).get(a, {}).items()
        for old in [base.get("outputs", {}).get(a, {}).get(name)]
        if old and new
    ]
    if not pairs:
        return done(
            "### Content diff\n\nNothing rebuilt; no outputs to compare.\n",
            {"pairs": 0, "identical": 0, "changed": 0, "skipped": 0},
        )

    identical, changed, skipped = compare_all(pairs)
    log(f"{len(pairs)} pairs: {len(identical)} identical, {len(changed)} changed, {len(skipped)} skipped")
    done(
        render(pairs, identical, changed, skipped),
        {
            "pairs": len(pairs),
            "identical": len(identical),
            "changed": len(changed),
            "skipped": len(skipped),
        },
    )


if __name__ == "__main__":
    main()

File: tools/wasinix/src/ci/contentdiff.rs
//! Whether rebuilt outputs actually changed content. A moved derivation says
//! only that something upstream changed; two builds can differ in every store
//! path and still be bit-identical once self-references are rewritten.

use std::collections::{BTreeMap, BTreeSet};

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::ci::evalmap::EvalMap;
use crate::support::atoms::JobAddr;
use crate::support::nix::Invocation;

const NARINFO_CHUNK: usize = 100;
/// Normalizing realises and rewrites both sides, which is far more expensive
/// than a hash comparison.
const MAX_NORMALIZE: usize = 25;
const NAR_SIZE_CAP: u64 = 256 * 1024 * 1024;

/// One output compared across the two sides.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Pair {
    pub attr: JobAddr,
    pub output: String,
    pub old: String,
    pub new: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Skipped {
    #[serde(flatten)]
    pub pair: Pair,
    pub reason: String,
}

/// The content comparison's outcome. It carries no status of its own: whether
/// the comparison ran to completion is the task's exit status, and a summary
/// only exists when it did.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContentSummary {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub identical: Vec<Pair>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub changed: Vec<Pair>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub skipped: Vec<Skipped>,
    /// Rebuilt jobs that failed to build, so they had no output to compare.
    #[serde(default, skip_serializing_if = "is_zero")]
    pub not_built: usize,
    /// Moved jobs excluded as content-free (validation checks).
    #[serde(default, skip_serializing_if = "is_zero")]
    pub excluded: usize,
}

fn is_zero(value: &usize) -> bool {
    *value == 0
}

impl ContentSummary {
    pub fn pair_count(&self) -> usize {
        self.identical.len() + self.changed.len() + self.skipped.len()
    }
}

fn basename(path: &str) -> &str {
    path.rsplit('/').next().unwrap_or(path)
}

/// Narinfo per store path, chunked and tolerant: nix path-info fails the
/// whole invocation when any queried path is absent, so a failed chunk
/// bisects until each missing path costs only itself, never its chunkmates.
fn path_infos(paths: &[String], store: Option<&str>) -> BTreeMap<String, Option<Value>> {
    fn query(
        chunk: &[String],
        store: Option<&str>,
        infos: &mut BTreeMap<String, Option<Value>>,
    ) {
        let mut invocation = Invocation::plain("path-info")
            .json()
            .args(["--json-format", "2"])
            .operands(chunk.iter().cloned());
        if let Some(store) = store {
            invocation = invocation.args(["--store", store]);
        }
        match invocation.probe("a missing path must not lose the comparison") {
            Ok(output) if output.status.is_success() => {
                if let Ok(value) = serde_json::from_slice::<Value>(&output.stdout) {
                    for (key, info) in value["info"].as_object().into_iter().flatten() {
                        infos.insert(
                            key.clone(),
                            if info.is_null() {
                                None
                            } else {
                                Some(info.clone())
                            },
                        );
                    }
                }
            }
            _ if chunk.len() > 1 => {
                let (left, right) = chunk.split_at(chunk.len() / 2);
                query(left, store, infos);
                query(right, store, infos);
            }
            result => {
                if let Ok(output) = result {
                    crate::support::ui::warning(format!(
                        "path-info: {}: {}",
                        basename(&chunk[0]),
                        output.stderr.trim().lines().last().unwrap_or_default()
                            .chars().take(160).collect::<String>()
                    ));
                }
                infos.insert(basename(&chunk[0]).to_string(), None);
            }
        }
    }
    let mut infos = BTreeMap::new();
    for chunk in paths.chunks(NARINFO_CHUNK) {
        query(chunk, store, &mut infos);
    }
    infos
}

/// Look locally first, then in the cache: a case built elsewhere has nothing
/// on this machine.
fn available_infos(paths: &[String], store: Option<&str>) -> BTreeMap<String, Option<Value>> {
    let mut local = path_infos(paths, None);
    let missing: Vec<String> = paths
        .iter()
        .filter(|path| local.get(basename(path)).is_none_or(Option::is_none))
        .cloned()
        .collect();
    if missing.is_empty() {
        return local;
    }
    let mut remote = path_infos(&missing, Some(store.unwrap_or(crate::support::nix::CACHE_SUBSTITUTER)));
    if store.is_some() {
        let still: Vec<String> = missing
            .iter()
            .filter(|path| remote.get(basename(path)).is_none_or(Option::is_none))
            .cloned()
            .collect();
        if !still.is_empty() {
            remote.extend(path_infos(&still, Some(crate::support::nix::CACHE_SUBSTITUTER)));
        }
    }
    for (key, value) in remote {
        if value.is_some() || !local.contains_key(&key) {
            local.insert(key, value);
        }
    }
    local
}

fn self_referential(path: &str, info: &Value) -> bool {
    info["references"]
        .as_array()
        .map(|refs| {
            refs.iter()
                .filter_map(Value::as_str)
                .any(|reference| basename(reference) == basename(path))
        })
        .unwrap_or(false)
}

/// Realise both sides and compare with self-references rewritten to content
/// hashes, which is the only way two self-referential outputs can be equal.
fn normalize_pair(old: &str, new: &str, store: Option<&str>) -> (Option<bool>, Option<String>) {
    let tail = |text: &str| {
        text.trim()
            .lines()
            .last()
            .unwrap_or_default()
            .chars()
            .take(120)
            .collect::<String>()
    };
    // Raw nix-store does not read the flake's nixConfig, so the cache is
    // named explicitly from the one place it is spelled.
    let realised = match store {
        Some(store) => Invocation::plain("copy")
            .args(["--from", store])
            .operands([old, new]),
        None => Invocation::tool("nix-store")
            .arg("--realise")
            .option(
                "extra-substituters",
                crate::support::nix::CACHE_SUBSTITUTER,
            )
            .option(
                "extra-trusted-public-keys",
                crate::support::nix::CACHE_PUBLIC_KEY,
            )
            .operands([old, new]),
    }
    .probe("a failed realise fails this pair, not the diff");
    match realised {
        Ok(output) if output.status.is_success() => {}
        Ok(output) => return (None, Some(format!("realise failed: {}", tail(&output.stderr)))),
        Err(error) => return (None, Some(format!("realise failed: {error}"))),
    }
    let output = match Invocation::plain("store make-content-addressed")
        .json()
        .operands([old, new])
        .probe("a failed normalize fails this pair, not the diff")
    {
        Ok(output) if output.status.is_success() => output,
        Ok(output) => {
            return (
                None,
                Some(format!(
                    "normalize failed: {}",
                    tail(&output.stderr)
                )),
            )
        }
        Err(error) => return (None, Some(format!("normalize failed: {error}"))),
    };
    let Ok(value) = serde_json::from_slice::<Value>(&output.stdout) else {
        return (None, Some("normalize produced no rewrites".into()));
    };
    let rewrites = &value["rewrites"];
    (Some(rewrites[old] == rewrites[new]), None)
}

fn compare_all(pairs: Vec<Pair>, store: Option<&str>) -> ContentSummary {
    let olds = available_infos(
        &pairs
            .iter()
            .map(|pair| pair.old.clone())
            .collect::<Vec<_>>(),
        store,
    );
    let news = available_infos(
        &pairs
            .iter()
            .map(|pair| pair.new.clone())
            .collect::<Vec<_>>(),
        store,
    );
    let mut result = ContentSummary::default();
    let mut normalized = 0;
    let skip = |result: &mut ContentSummary, pair: &Pair, reason: String| {
        result.skipped.push(Skipped {
            pair: pair.clone(),
            reason,
        });
    };

    for pair in &pairs {
        let old_info = olds.get(basename(&pair.old)).and_then(Option::as_ref);
        let new_info = news.get(basename(&pair.new)).and_then(Option::as_ref);
        if pair.old == pair.new {
            result.identical.push(pair.clone());
            continue;
        }
        let (Some(old_info), Some(new_info)) = (old_info, new_info) else {
            let side = if old_info.is_none() { "base" } else { "new" };
            skip(&mut result, pair, format!("{side} output not available"));
            continue;
        };
        if old_info["narHash"] == new_info["narHash"] {
            result.identical.push(pair.clone());
        } else if !self_referential(&pair.old, old_info) && !self_referential(&pair.new, new_info)
        {
            result.changed.push(pair.clone());
        } else if old_info["narSize"]
            .as_u64()
            .unwrap_or(0)
            .max(new_info["narSize"].as_u64().unwrap_or(0))
            > NAR_SIZE_CAP
        {
            skip(
                &mut result,
                pair,
                "self-referential and too large to normalize".into(),
            );
        } else if normalized >= MAX_NORMALIZE {
            skip(
                &mut result,
                pair,
                format!("normalize cap ({MAX_NORMALIZE}) reached"),
            );
        } else {
            normalized += 1;
            match normalize_pair(&pair.old, &pair.new, store) {
                (Some(true), _) => result.identical.push(pair.clone()),
                (Some(false), _) => result.changed.push(pair.clone()),
                (None, reason) => skip(&mut result, pair, reason.unwrap_or_default()),
            }
        }
    }
    result
}

/// Jobs whose derivation moved and whose outputs are worth comparing.
/// Validation checks emit success tokens, which carry no content.
pub fn content_jobs(
    base: &EvalMap,
    head: &EvalMap,
    allowed: Option<&BTreeSet<String>>,
) -> (Vec<String>, usize) {
    let moved: Vec<String> = head
        .jobs
        .iter()
        .filter(|(attr, drv)| {
            base.jobs.get(attr.as_str()).is_some_and(|old| old != *drv)
                && allowed.is_none_or(|allowed| allowed.contains(attr.as_str()))
        })
        .map(|(attr, _)| attr.as_str().to_string())
        .collect();
    let included: Vec<String> = moved
        .iter()
        .filter(|attr| {
            head.info
                .get(attr.as_str())
                .map(|info| info.content_diff)
                .unwrap_or_else(|| !attr.starts_with("checks."))
        })
        .cloned()
        .collect();
    let excluded = moved.len() - included.len();
    (included, excluded)
}

/// The output pairs a run can compare: every shared output of the jobs that
/// moved, built on both sides.
pub fn pairs_of(
    base: &EvalMap,
    head: &EvalMap,
    jobs: &[String],
    failed: &BTreeSet<String>,
) -> Vec<Pair> {
    jobs.iter()
        .filter(|attr| !failed.contains(attr.as_str()))
        .flat_map(|attr| {
            let old_outputs = base.outputs.get(attr.as_str()).cloned().unwrap_or_default();
            let new_outputs = head.outputs.get(attr.as_str()).cloned().unwrap_or_default();
            new_outputs
                .into_iter()
                .filter_map(|(name, new)| {
                    let old = old_outputs.get(&name)?.clone();
                    (!old.is_empty() && !new.is_empty()).then_some(Pair {
                        attr: JobAddr(attr.clone()),
                        output: name,
                        old,
                        new,
                    })
                })
                .collect::<Vec<_>>()
        })
        .collect()
}

pub struct Request<'a> {
    pub base_map: &'a EvalMap,
    pub head_map: &'a EvalMap,
    pub junit: &'a [std::path::PathBuf],
    pub allowed_jobs: Option<&'a BTreeSet<String>>,
    /// Probe this store before the public cache, which is where a
    /// store-routed build left its outputs.
    pub store: Option<&'a str>,
}

/// Attrs whose build or evaluation failed have no new output to compare. An
/// upload failure still produced one.
fn failed_attrs(junit: &[std::path::PathBuf]) -> BTreeSet<String> {
    crate::ci::compare::junit_status(junit)
        .into_iter()
        .filter(|(_, status)| *status == crate::support::atoms::JobStatus::Failure)
        .map(|(attr, _)| attr.0)
        .collect()
}

pub fn run(request: &Request<'_>) -> ContentSummary {
    let (jobs, excluded) = content_jobs(request.base_map, request.head_map, request.allowed_jobs);
    let failed = failed_attrs(request.junit);
    let not_built = jobs.iter().filter(|attr| failed.contains(attr.as_str())).count();
    let pairs = pairs_of(request.base_map, request.head_map, &jobs, &failed);
    let mut summary = if pairs.is_empty() {
        ContentSummary::default()
    } else {
        compare_all(pairs, request.store)
    };
    summary.not_built = not_built;
    summary.excluded = excluded;
    summary
}
